src/accountsdb/download.zig

```diff
@@ -0,0 +1,652 @@
+                self.bytes_read * 100 / self.download_size,
+                mb_per_second,
+                std.fmt.fmtDuration(time_left_ns),
+            });
+
+            const should_check_speed = self.min_mb_per_second != null and !self.has_checked_speed;
+            if (should_check_speed) {
+                // dont check again
+                self.has_checked_speed = true;
+                if (mb_per_second < self.min_mb_per_second.?) {
+                    // not fast enough => abort
+                    self.logger.infof("download speed is too slow ({d} MB/s) -- disconnecting", .{mb_per_second});
+                    return 0;
+                } else {
+                    self.logger.infof("download speed is ok ({d} MB/s) -- maintaining connection\r", .{mb_per_second});
+                }
+            }
+        }
+
+        return len;
+    }
+};
+
+fn checkCode(code: curl.libcurl.CURLcode) !void {
+    if (code == curl.libcurl.CURLE_OK) {
+        return;
+    }
+    // https://curl.se/libcurl/c/libcurl-errors.html
+    std.log.debug("curl err code:{d}, msg:{s}\n", .{ code, curl.libcurl.curl_easy_strerror(code) });
+    return error.Unexpected;
+}
+
+pub fn setNoBody(self: curl.Easy, no_body: bool) !void {
+    try checkCode(curl.libcurl.curl_easy_setopt(self.handle, curl.libcurl.CURLOPT_NOBODY, @as(c_long, @intFromBool(no_body))));
+}
+
+pub fn downloadFile(
+    allocator: std.mem.Allocator,
+    logger: Logger,
+    url: [:0]const u8,
+    output_dir: []const u8,
+    filename: []const u8,
+    min_mb_per_second: ?usize,
+) !bool {
+    var easy = try curl.Easy.init(allocator, .{});
+    defer easy.deinit();
+
+    try easy.setUrl(url);
+    try easy.setMethod(.HEAD);
+    try setNoBody(easy, true);
+    var head_resp = easy.perform() catch {
+        return error.HeaderRequestFailed;
+    };
+
+    var download_size: usize = 0;
+    if (try head_resp.getHeader("content-length")) |content_length| {
+        download_size = try std.fmt.parseInt(usize, content_length.get(), 10);
+    } else {
+        logger.debugf("header request didnt have content-length...", .{});
+        return error.NoContentLength;
+    }
+
+    // timeout will need to be larger
+    easy.timeout_ms = std.time.ms_per_hour * 5; // 5 hours is probs too long but its ok
+    var download_progress = try DownloadProgress.init(
+        logger,
+        output_dir,
+        filename,
+        download_size,
+        min_mb_per_second,
+    );
+
+    try setNoBody(easy, false); // full download
+    try easy.setUrl(url);
+    try easy.setMethod(.GET);
+    try easy.setWritedata(&download_progress);
+    try easy.setWritefunction(DownloadProgress.bufferWriteCallback);
+
+    var resp = easy.perform() catch return false;
+    defer resp.deinit();
+
+    const full_download = download_progress.file_memory_index == download_size;
+    if (!full_download) {
+        // too slow = early exit
+        return false;
```

Comment:
This feels like an error. Is it conceivable that this would be acceptable? For example, maybe the download can be resumed, or the partial file could be useful somehow.

Reply:
agreed error is more accurate - 507e95b